src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -556,6 +561,44 @@ where `<compatible-ver>` is the latest version supporting rustc {rustc_version}"
         }
     }
 
+    if build_config.detect_antivirus != DetectAntivirus::Never {
+        // Count the number of test binaries and build scripts we'll need to
+        // run. This doesn't take into account the binary that will be run
+        // if `cargo run` was specified, and doesn't handle pre-2024 `rustdoc`
+        // tests, but that's fine, this is only a heuristic.
+        let num_binaries = unit_graph
+            .keys()
+            .filter(|unit| {
+                matches!(
+                    unit.mode,
+                    CompileMode::Test | CompileMode::Doctest | CompileMode::RunCustomBuild
+                )
+            })
+            .count();
+
+        tracing::debug!("estimated {num_binaries} binaries that could be slowed down by antivirus");
+
+        // TODO(madsmtm): Consider only warning once every X days.
+
+        // Heuristic: Only do the check if we have to run more than a specific
+        // number of binaries. This makes it so that small beginner projects
+        // don't hit this.
+        //
+        // TODO(madsmtm): What should the threshold be?
+        //
+        // We also don't want to do this check when installing, since there
+        // might be `cargo install` users who are not necessarily developers
+        // (and so the note will be irrelevant to them).
+        if (10 < num_binaries && build_config.intent != UserIntent::Install)
+            || build_config.detect_antivirus == DetectAntivirus::Always
+        {
+            if let Err(err) = detect_antivirus::detect_and_report(gctx) {
```

Comment:
One thing I wanted to note is that there is a performance cost to this. At least on my system, an error result took over 60ms, and a passing result took 14ms. We try to keep cargo's overhead relatively small (I like to keep the budget under 500ms total). 14ms is probably ok, but it's inching towards the territory where it is significant.